### DISCUSSION TITLE
RetryWhen with max retry count

### DISCUSSION THREAD

**Discussion by softlion** [dotnet/reactive#1812] [Ideas]
Retry has a retryCount option. But RetryWhen does not. It would be nice to have the same option in RetryWhen. And would make the Retry* methods more consistent.

**Comment by akarnokd** [comment]
With `RetryWhen` you are in full control of when to retry via the inner sequence. Just count the errors received from the error sequence.

https://github.com/dotnet/reactive/blob/9f2a8090cea4bf931d4ac3ad071f4df147f4df50/Rx.NET/Source/tests/Tests.System.Reactive/Tests/Linq/Observable/RetryWhenTest.cs#L130-L178

**Reply by softlion** [reply]
" Just count the errors received from the error sequence."

Sorry 1) I don't have the level to find/understand that and 2) a helper version doing that would be much simpler for everyone as this is a common requested feature.

Or maybe the philosophy here is to exclude all developpers having hard time to keep rx confidential. I don't understand.

**Reply by akarnokd** [reply]
There is nothing confidential about Rx. You simply need to work with existing tools most of the time.

Not sure what the problem is with counting the errors. I provided the example that does error counting above. Here is an excerpt:

```cs
source.RetryWhen(errors => 
         { 
             int[] count = { 0 }; 
             return errors.SelectMany(error => 
             { 
                 var c = ++count[0]; 
                 if (c == 3) 
                 { 
                     return Observable.Throw<int>(error); 
                 } 
                 return Observable.Return(1); 
             }); 
         }) 
```

Make this an extension method in your project with callbacks as you see fit.
